// automation_proj/pages/loginPage.ts
import { Page, Locator } from '@playwright/test';

export class LoginPage {
  private page: Page;
  private loginButton: Locator;
  private emailInput: Locator;
  private nextButton: Locator;
  private passwordInput: Locator;
  private signInButton: Locator;
  private noButton: Locator;
  private closeButton: Locator;

  constructor(page: Page) {
    this.page = page;
    this.loginButton = page.locator('button:has-text("LOGIN")');
    this.emailInput = page.getByRole('textbox', { name: 'Enter your email, phone, or' });
    this.nextButton = page.locator('input[type="submit"][value="Next"]');
    this.passwordInput = page.locator('input[type="password"][name="passwd"]');
    this.signInButton = page.locator('input[type="submit"][value="Sign in"]');
    this.noButton = page.getByRole('button', { name: 'No' });
    this.closeButton = page.getByRole('button', { name: 'Close' });
  }

  async navigate() {
    await this.page.goto('https://computrace7test.thermon-rd.com/', { waitUntil: 'domcontentloaded' });
    await this.loginButton.waitFor({ state: 'visible', timeout: 15000 }); // Esperar a que el botón de LOGIN esté visible
  }

  async login(email: string, password: string) {
    // Hacer clic en el botón de LOGIN
    await this.loginButton.click();

    // Esperar y llenar el campo de email
    await this.emailInput.waitFor({ state: 'visible', timeout: 15000 });
    await this.emailInput.fill(email);

    // Hacer clic en el botón "Next"
    await this.nextButton.click();

    // Esperar y llenar el campo de contraseña
    await this.passwordInput.waitFor({ state: 'visible', timeout: 15000 });
    await this.passwordInput.fill(password);

    // Hacer clic en el botón "Sign in"
    await this.signInButton.click();

    // Esperar y hacer clic en el botón "No" (si está presente)
    await this.noButton.waitFor({ state: 'visible', timeout: 15000 });
    await this.noButton.click();

    // Esperar y hacer clic en el botón "Close" (si está presente)
    await this.closeButton.waitFor({ state: 'visible', timeout: 15000 });
    await this.closeButton.click();
  }
}
